### PR TITLE
dwarf: fixes for non-exact matched lines 

### PR DESCRIPTION
**Pull request by mathetake**

### CODE REVIEW

File: internal/wasmdebug/dwarf.go [hunk 1]
```diff
@@ -137,44 +137,68 @@ entry:
 		return
 	}
 
+	ln := lines[index]
+	if ln.addr != instructionOffset {
+		// If the address doesn't match exactly, the previous entry is the one that contains the instruction.
+		if index-1 < 0 {
+			return
+		}
+		ln = lines[index-1]
```
Comment (mathetake) [inline internal/wasmdebug/dwarf.go:146]: this is the actual fix and others are just cosmetic

**Comment by codefromthecrypt** [comment]
I think there are some wasm tools out there that allow you to snip certain sections.. I wonder if we can get that rust wasm smaller as it is nearly 2MB compared to the tinygo dwarf which is 40k

File: internal/wasmdebug/dwarf.go [hunk 2]
```diff
@@ -137,44 +137,68 @@ entry:
 		return
 	}
 
+	ln := lines[index]
+	if ln.addr != instructionOffset {
+		// If the address doesn't match exactly, the previous entry is the one that contains the instruction.
+		if index-1 < 0 {
```
Comment (codefromthecrypt) [inline internal/wasmdebug/dwarf.go:143]: maybe add a "for example" here with a high level of how something doesn't match exactly (don't need to paste the trace, just mention high level about the rust maybe with inlining which can cause this)

Comment (mathetake) [reply internal/wasmdebug/dwarf.go:146]: https://github.com/tetratelabs/wazero/pull/916/commits/cee9b05123a2cb22ad8d06548e985c8a1d81f461 added comment

**Comment by mathetake** [comment]
so decided to do gzipping :D

**Comment by mathetake** [comment]
ok now `494 KB`

**Comment by mathetake** [comment]
```
-rw-r--r--  1 mathetake  staff   494K Dec 13 12:17 main.wasm.gz
-rw-r--r--  1 mathetake  staff   372K Dec 13 11:18 main.wasm.xz
``` 

xz is better

**Comment by mathetake** [comment]
ok done with xz

**Comment by mathetake** [comment]
finally all green

File: Makefile [hunk 3]
```diff
@@ -107,7 +107,13 @@ build.examples.emscripten: $(emscripten_sources)
...
 
 .PHONY: build.examples.rust
-build.examples.rust: examples/allocation/rust/testdata/greet.wasm imports/wasi_snapshot_preview1/example/testdata/cargo-wasi/cat.wasm imports/wasi_snapshot_preview1/testdata/cargo-wasi/wasi.wasm
+build.examples.rust: examples/allocation/rust/testdata/greet.wasm imports/wasi_snapshot_preview1/example/testdata/cargo-wasi/cat.wasm imports/wasi_snapshot_preview1/testdata/cargo-wasi/wasi.wasm internal/testing/dwarftestdata/testdata/rust/main.wasm.xz
+
+# Normally, we build release because it is smaller. Testing dwarf requires the debug build.
+internal/testing/dwarftestdata/testdata/rust/main.wasm.xz:
+	cd $(@D) && cargo wasi build
+	mv $(@D)/target/wasm32-wasi/debug/main.wasm $(@D)
+	cd $(@D) && xz -k -f ./main.wasm # Rust's DWARF section is huge, so compress as gzip.
```
Comment (codefromthecrypt) [inline Makefile:116]:
```suggestion
	cd $(@D) && xz -k -f ./main.wasm # Rust's DWARF section is huge, so compress it.
```

**Review by codefromthecrypt — APPROVED** [review]
great job on the code and the infra. <400kb is better than almost 2MB and keeps the engineering simpler

**Pull request merged**